### ISSUE TITLE
Catch type-related server errors

### ISSUE THREAD

**Issue by st31ny** [bcb/jsonrpcserver#119]
Due to a bug in my code, a method returned a non-JSON-encodable object. This resulted in a `TypeError` caused by the repsose logging:
```
   File "/usr/lib/python3/dist-packages/jsonrpcserver/async_dispatcher.py", line 103, in dispatch
    log_response(str(response), trim_log_values=trim_log_values)
  File "/usr/lib/python3/dist-packages/jsonrpcserver/response.py", line 133, in __str__
    return json.dumps(sort_dict_response(self.deserialized()))
  File "/usr/lib/python3.8/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python3.8/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python3.8/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib/python3.8/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type bytes is not JSON serializable
```

Instead of failing `dispatch` completely, I'd rather expect just the call to fail with an internal server error sent to the client (which holds especially true in the case of batch requests where only one of the requests fails).

For this issue, I'd propose to add a check in the `SuccessResponse.__init__` function that checks if the passed object is actually JSON-encodable. Alternatively, `safe_call()` could call `str()` on the response after creation within the `handle_exception` block.

I'd be happy to contribute once we've settled on a solution.

**Comment by bcb** [comment]
My suggestion would be at the end of dispatch_pure, if it the value's not json-serializable return an `ErrorResponse` with code -32000 (server error) and http status 500.

Would need to be done in both sync and async versions of dispatch_pure.

**Comment by bcb** [comment]
Actually you're right, I think inside safe_call is the right place.

**Comment by bcb** [comment]
Fixed in 4.1.3.
